src/diracx/routers/auth.py

```diff
@@ -648,84 +711,212 @@ async def token(
+    redirect_uri: str | None,
+    code_verifier: str | None,
+    auth_db: AuthDB,
+    settings: AuthSettings,
+):
+    """Get OIDC token information from the authorization flow DB and check few parameters before returning it"""
+    assert code is not None
+    info = await auth_db.get_authorization_flow(
+        code, settings.authorization_flow_expiration_seconds
     )
+    if redirect_uri != info["redirect_uri"]:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Invalid redirect_uri",
+        )
+
+    try:
+        assert code_verifier is not None
+        code_challenge = (
+            base64.urlsafe_b64encode(hashlib.sha256(code_verifier.encode()).digest())
+            .decode()
+            .strip("=")
+        )
+    except Exception as e:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Malformed code_verifier",
+        ) from e
+
+    if code_challenge != info["code_challenge"]:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Invalid code_challenge",
+        )
+
+    oidc_token_info = info["id_token"]
+    scope = info["scope"]
+
+    # TODO: use HTTPException while still respecting the standard format
+    # required by the RFC
+    if info["status"] != FlowStatus.READY:
+        # That should never ever happen
+        raise NotImplementedError(f"Unexpected flow status {info['status']!r}")
+
+    return (oidc_token_info, scope)
+
+
+async def get_oidc_token_info_from_refresh_flow(
+    refresh_token: str | None, auth_db: AuthDB, settings: AuthSettings
+):
+    """Get OIDC token information from the refresh token DB and check few parameters before returning it"""
+    assert refresh_token is not None
+
+    # Decode the refresh token to get the JWT ID
+    jti, _ = await verify_dirac_refresh_token(refresh_token, settings)
+
+    # Get some useful user information from the refresh token entry in the DB
+    refresh_token_attributes = await auth_db.get_refresh_token(jti)
+
+    sub = refresh_token_attributes["sub"]
+
+    # Refresh token rotation: https://datatracker.ietf.org/doc/html/rfc6749#section-10.4
+    # Check that the refresh token has not been already revoked
+    # This might indicate that a potential attacker try to impersonate someone
+    # In such case, all the refresh tokens bound to a given user (subject) should be revoked
+    # Forcing the user to reauthenticate interactively through an authorization/device flow (recommended practice)
+    if refresh_token_attributes["status"] == RefreshTokenStatus.REVOKED:
+        # Get all the user tokens from the subject
+        res = await auth_db.get_user_refresh_tokens(sub)
+
+        # Revoke each of them
+        for refresh_token_to_revoke in res:
+            await auth_db.revoke_refresh_token(refresh_token_to_revoke["jti"])
+            # TODO: what if revoke_refresh_token() raises an exception?
```

Comment:
I'm not sure we can do anything about this, though having it in a single SQL statement will hopefully make it more robust.